### PR TITLE
Improve precision of which transaction caused balance to go negative

### PR DESCRIPTION
**Pull request by qwhelan**
The existing logic accumulates all buys first before decrementing from sells. As a result, it can only detect negative balances that persist until the end of the calculation period. It also means that the blamed transaction can be incorrect. Consider the following transaction sequence:
```
BUY 0.5 BTC
SELL 0.9 BTC
BUY 0.5 BTC
SELL 0.2 BTC
```
Under the current logic, an exception would be raised when processing the `SELL 0.2 BTC` transaction. This code, would raise an exception on the `SELL 0.9 BTC` transaction instead. A subtler case looks like this:
```
BUY 0.5 BTC
SELL 0.9 BTC
BUY 0.5 BTC
SELL 0.1 BTC
```
The current code would not raise any exception, despite going significantly negative for a period. The new code would correctly flag the invariant being violated.

### CODE REVIEW

**Comment by eprbell** [comment]
If you like you can also run checks locally on your machine with `make static_analysis check`.

**Comment by qwhelan** [comment]
@eprbell Thanks, static analysis should be passing.

Completely forgot to mention this when opening the PR - a lot of test data fails this check now, which leads to a bunch of failing tests.

Easiest way would be to turn the check off for these tests, but there's a chance other issues are being hidden by the test data being off. For example, tab `B2` of `input/test_data.ods` now has this failure in `tests/test_tax_engine.py::TestTaxEngine::test_good_input`:
```
E                   rp2.rp2_error.RP2ValueError: B2 balance of account "Coinbase" (holder "Bob") went negative (-0.20000000000) on the following transaction: OutTransaction:
E                     id=15
E                     timestamp=2020-02-11 19:58:00.000000 +0000
E                     asset=B2
E                     exchange=Coinbase
E                     holder=Bob
E                     transaction_type=TransactionType.SELL
E                     spot_price=12200.0000
E                     crypto_out_no_fee=1.00000000
E                     crypto_fee=0.00000000
E                     unique_id=
E                     is_taxable=True
E                     fiat_taxable_amount=12200.0000
```
Relevant cells showing the error to be correct are highlighted in yellow:
Screenshot from 2024-05-08 21-09-55

**Comment by eprbell** [comment]
I still need to look at the code of this PR (I'll do that in the weekend or sooner), but this is a good check, so thanks for identifying and fixing this problem! 

In fact I had already found that some tests end up with a negative balance (even without this fix) and added the `-n` switch to disable the check for negative balances: see https://github.com/eprbell/rp2/commit/195ed40a057550231c586f669505431bdf2438df. 

I imagine this PR probably causes even more tests balances to go negative, however I'm reluctant to change the tests: the reason is that I did a lot of manual verification of these tests and they estabilish a strong baseline against regression (the balance may go negative here and there, but the tax engine logic is still tested fully and thoroughly by these tests).

So my suggestion is to:
* disable negative balance checks (only for the tests that run into the issue), similar to what happens here: https://github.com/eprbell/rp2/blob/main/tests/test_ods_output_diff.py#L37
* write new tests running with this check enabled and testing negative balances specifically (I think you already added some of these tests). 

This should give us the best of both worlds. Do you have any concerns with this approach?

**Comment by qwhelan** [comment]
Makes sense as an approach - my main concern is having a giant PR that's infeasible to work with. Hopefully it's feasible to land smaller bits using your `-n` switch to keep things tractable. I'm going to be out of town this weekend, so might be a bit before I take a stab at things (and might submit some other unrelated PRs first).

Just for an order-of-magnitude, I'm seeing 47 tests passing, 1 failed, and 66 errors. The latter bucket seems to all be ODS output diff style tests.

File: src/rp2/balance.py [hunk 1]
```diff
@@ -119,53 +118,60 @@ def __init__(
...
-                and not configuration.allow_negative_balances
-            ):
-                raise RP2ValueError(
-                    f'{intra_transaction.asset} balance of account "{from_account.exchange}" (holder "{from_account.holder}") went negative '
-                    f'({final_balances[from_account]}) on the following transaction: {intra_transaction}'
-                )
-
-        # Balances for sold and gifted currency
-        for transaction in self.__input_data.unfiltered_out_transaction_set:
+        transactions: List[Union[InTransaction, IntraTransaction, OutTransaction]] = in_transactions + intra_transactions + out_transactions
```
Comment (eprbell) [inline src/rp2/balance.py:125]: Could this be `List[AbstractTransaction]` instead of a `Union`?

**Review by eprbell — COMMENTED** [review]
I did a first review: looks good. After we fix the tests, as discussed above, this can be merged.

File: src/rp2/balance.py (hunk 1 above)
Comment (qwhelan) [reply src/rp2/balance.py]: Solved this in the most recent commit via this approach: https://peps.python.org/pep-0484/#type-variables-with-an-upper-bound.

This can be simplified when 3.11 is the minimum supported version via PEP 673

File: src/rp2/abstract_entry_set.py [hunk 2]
```diff
@@ -24,6 +24,8 @@
 from rp2.out_transaction import OutTransaction
 from rp2.rp2_error import RP2TypeError, RP2ValueError
 
+T = TypeVar("T", bound="AbstractEntrySet")
```
Comment (eprbell) [inline src/rp2/abstract_entry_set.py:27]: Can you use a descriptive name for `T` (something like `AbstractEntrySetSubclass`)?

Comment (qwhelan) [reply src/rp2/abstract_entry_set.py]: Updated

File: tests/test_balance.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
+# Copyright 2021 eprbell
```
Comment (eprbell) [inline tests/test_balance.py:1]: The convention we've been following is that new files have the creator in the copyright line: 
```suggestion
# Copyright 2024 qwhelan
```

File: tests/test_tax_engine.py [hunk 4]
```diff
@@ -33,12 +33,14 @@
 
 class TestTaxEngine(unittest.TestCase):
     _good_input_configuration: Configuration
+    _good_input_allow_negative_configuration: Configuration
```
Comment (eprbell) [inline tests/test_tax_engine.py:36]: Let's use a fully spelled-out name (it will make it easier to understand it's purpose):
```suggestion
    _good_input_allow_negative_balance_configuration: Configuration
```

**Review by eprbell — CHANGES_REQUESTED** [review]
This looks great: I have just a couple of nits and then we can merge this.

File: tests/test_balance.py (hunk 3 above)
Comment (qwhelan) [reply tests/test_balance.py]: Done

**Review by eprbell — APPROVED** [review]
LGTM. Thanks for fixing this!
